contrib/adaptive-compression/adapt.c

```diff
@@ -0,0 +1,591 @@
+static unsigned g_displayStats = 0;
+static UTIL_time_t g_startTime;
+static size_t g_streamedSize = 0;
+static unsigned g_useProgressBar = 0;
+static UTIL_freq_t g_ticksPerSecond;
+
+typedef struct {
+    void* start;
+    size_t size;
+} buffer_t;
+
+typedef struct {
+    unsigned waitCompressed;
+    unsigned waitReady;
+    unsigned waitWrite;
+    unsigned readyCounter;
+    unsigned compressedCounter;
+    unsigned writeCounter;
+} cStat_t;
+
+typedef struct {
+    buffer_t src;
+    buffer_t dst;
+    unsigned compressionLevel;
+    unsigned jobID;
+    size_t compressedSize;
+} jobDescription;
+
+typedef struct {
+    unsigned compressionLevel;
+    unsigned numActiveThreads;
+    unsigned numJobs;
+    unsigned lastJobID;
+    unsigned nextJobID;
+    unsigned threadError;
+    unsigned jobReadyID;
+    unsigned jobCompressedID;
+    unsigned jobWriteID;
+    unsigned allJobsCompleted;
+    unsigned adaptParam;
+    pthread_mutex_t jobCompressed_mutex;
+    pthread_cond_t jobCompressed_cond;
+    pthread_mutex_t jobReady_mutex;
+    pthread_cond_t jobReady_cond;
+    pthread_mutex_t allJobsCompleted_mutex;
+    pthread_cond_t allJobsCompleted_cond;
+    pthread_mutex_t jobWrite_mutex;
+    pthread_cond_t jobWrite_cond;
+    cStat_t stats;
+    jobDescription* jobs;
+    FILE* dstFile;
+    ZSTD_CCtx* cctx;
+} adaptCCtx;
+
+static void freeCompressionJobs(adaptCCtx* ctx)
+{
+    unsigned u;
+    for (u=0; u<ctx->numJobs; u++) {
+        jobDescription job = ctx->jobs[u];
+        free(job.dst.start);
+        free(job.src.start);
+    }
+}
+
+static int freeCCtx(adaptCCtx* ctx)
+{
+    {
+        int const compressedMutexError = pthread_mutex_destroy(&ctx->jobCompressed_mutex);
+        int const compressedCondError = pthread_cond_destroy(&ctx->jobCompressed_cond);
+        int const readyMutexError = pthread_mutex_destroy(&ctx->jobReady_mutex);
+        int const readyCondError = pthread_cond_destroy(&ctx->jobReady_cond);
+        int const allJobsMutexError = pthread_mutex_destroy(&ctx->allJobsCompleted_mutex);
+        int const allJobsCondError = pthread_cond_destroy(&ctx->allJobsCompleted_cond);
+        int const jobWriteMutexError = pthread_mutex_destroy(&ctx->jobWrite_mutex);
+        int const jobWriteCondError = pthread_cond_destroy(&ctx->jobWrite_cond);
+        int const fileCloseError =  (ctx->dstFile != NULL && ctx->dstFile != stdout) ? fclose(ctx->dstFile) : 0;
+        int const cctxError = ZSTD_isError(ZSTD_freeCCtx(ctx->cctx)) ? 1 : 0;
+        if (ctx->jobs){
+            freeCompressionJobs(ctx);
+            free(ctx->jobs);
+        }
+        return compressedMutexError | compressedCondError | readyMutexError | readyCondError | fileCloseError | allJobsMutexError | allJobsCondError | jobWriteMutexError | jobWriteCondError | cctxError;
+    }
+}
+
+static adaptCCtx* createCCtx(unsigned numJobs, const char* const outFilename)
+{
+
+    adaptCCtx* ctx = malloc(sizeof(adaptCCtx));
```

Comment:
Use `calloc()` instead, it is more efficient and standard than `malloc()` followed by `memset()`.